src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,356 @@
+                    s.finish() as usize
+                }) % classes.len()];
+                imp.sender_label.add_css_class(color_class);
+
+                *sender_color_class = Some(color_class.into());
+
+                imp.sender_label.set_visible(true);
+            } else {
+                imp.sender_label.set_visible(false);
+            }
+
+            self.update_document(message);
+        } else {
+            unreachable!("Unexpected message type: {:?}", message);
+        }
+
+        imp.message.replace(Some(message));
+        self.notify("message");
+    }
+}
+
+#[derive(PartialEq)]
+enum FileStatus {
+    Downloading(f64),
+    Uploading(f64),
+    CanBeDownloaded,
+    Downloaded,
+}
+use FileStatus::*;
+
+impl From<&File> for FileStatus {
+    fn from(file: &File) -> Self {
+        let local = &file.local;
+        let remote = &file.remote;
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        if local.is_downloading_active {
+            let progress = local.downloaded_size as f64 / size as f64;
+            Downloading(progress)
+        } else if remote.is_uploading_active {
+            let progress = remote.uploaded_size as f64 / size as f64;
+            Uploading(progress)
+        } else if local.is_downloading_completed {
+            Downloaded
+        } else if local.can_be_downloaded {
+            CanBeDownloaded
+        } else {
+            dbg!(file);
+            unimplemented!("unknown file status");
+        }
+    }
+}
+
+impl MessageDocument {
+    fn update_document(&self, message: &Message) {
+        if let MessageContent::MessageDocument(data) = message.content().0 {
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name_label.set_label(&data.document.file_name);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, file: File, session: Session) -> FileStatus {
+        let status = FileStatus::from(&file);
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        self.update_size_label(&status, size);
+        self.update_button(file, session, &status);
+
+        status
+    }
+
+    fn block_click(&self, block: bool) {
+        let imp = self.imp();
+        let click = &*imp.click;
+        let document_box = &*imp.document_box;
+
+        if let Some(ref handler_id) = *imp.click_handler_id.borrow() {
+            if block {
+                click.block_signal(handler_id);
+                document_box.remove_css_class("activatable");
+            } else {
+                click.unblock_signal(handler_id);
+                document_box.add_css_class("activatable");
+            }
+        }
+    }
```

Comment:
I don't see any freezing, you just can't stop dowloading by clicking on document

[Kooha-2022-09-11-15-36-37.webm](https://user-images.githubusercontent.com/63008755/189527885-f1c8ba11-f556-4d89-b522-179a03978319.webm)

Just like in any other telegram client

https://user-images.githubusercontent.com/63008755/189528090-0ab5573c-fce0-4452-a6e0-a3f1afc279f4.mp4

Reply:
Yeah, maybe I was too quick with my conclusion. This probably won't cause the freeze. But it sometimes freezes and we need to investigate why.


https://user-images.githubusercontent.com/3630213/189529115-fdc96cb9-ae5b-432b-bafe-5965f382c1e2.mp4

Reply:
I found that telegrand can freeze on scrolling because of too much calls of MessageLabel::measure

[telegrand-flamegraph](https://user-images.githubusercontent.com/63008755/189529347-af677541-a629-4310-998f-254574460454.svg) (open flamegraph with browser to be able zoom)

But it isn't related to document

Reply:
> I found that telegrand can freeze on scrolling because of too much calls of MessageLabel::measure

Yep that's known, it's a GTK performance issue, we definitely need to report this and help fixing it at some point.

Reply:
@YuraIz Btw thanks for the flamegraph, can you open an issue about that? This way we can discuss it separately.

Reply:
Ok, probably we are running in the same issue here.

Reply:
It's still really weird though in that case specifically. The issue about low performance GtkLabels should happen only with long multi-line text, and only in the measure phase (so when you update the text itself).

@marhkb does the program ever function again after that freeze? Unfortunately I'm out of town so I can't test this right now (and probably for a few days).

Reply:
@melix99 Interesting idea. I will check this in the evening.